### PR TITLE
feature#1117-export recommender model

### PR DESCRIPTION
**Pull request by clarkhywh**
1. We modified the front end in RecommenderPanel.html and added a download button.
2. We have modified the initialization function in RecommenderInfoPanel.java.
3.RecommendationEngine added a new method to export recommendation objects
4. We implemented the export function in RecommendationEngine.java in StringMatchingRecommender.java
5. There are many subclasses or related classes that also need to implement the export function in RecommendationEngine.java
6. We don't know how to convert the format to the output stream available to java.

### CODE REVIEW

File: inception-imls-stringmatch/src/main/java/de/tudarmstadt/ukp/inception/recommendation/imls/stringmatch/StringMatchingRecommender.java [hunk 1]
```diff
@@ -84,6 +107,53 @@ public StringMatchingRecommender(Recommender aRecommender,
         traits = aTraits;
         gazeteerService = aGazeteerService;
     }
+    
+    @Override
+    public void export(RecommenderContext aContext, OutputStream aTarget) throws IOException
```
Comment (munterkalmsteiner) [inline inception-imls-stringmatch/src/main/java/de/tudarmstadt/ukp/inception/recommendation/imls/stringmatch/StringMatchingRecommender.java]: Where is this method called?

File: inception-imls-stringmatch/src/main/java/de/tudarmstadt/ukp/inception/recommendation/imls/stringmatch/StringMatchingRecommender.java [hunk 2]
```diff
@@ -84,6 +107,53 @@ public StringMatchingRecommender(Recommender aRecommender,
...
+
+        User user = userService.getCurrentUser();
+
+        CAS cas = new CASImpl();
+
+        List<CAS> casList=new ArrayList<>();
+        casList.add(cas);
+
+        try {
+            train(aContext, casList );
```
Comment (munterkalmsteiner) [inline inception-imls-stringmatch/src/main/java/de/tudarmstadt/ukp/inception/recommendation/imls/stringmatch/StringMatchingRecommender.java]: Why is training needed for exporting the model?

File: inception-imls-stringmatch/src/main/java/de/tudarmstadt/ukp/inception/recommendation/imls/stringmatch/StringMatchingRecommender.java [hunk 3]
```diff
@@ -84,6 +107,53 @@ public StringMatchingRecommender(Recommender aRecommender,
...
+        // Extract all predictions for the current document / recommender
+        List<AnnotationSuggestion> suggestions = predictions.getPredictions().entrySet().stream()
+                .filter(f -> f.getKey().getDocumentName().equals(sourceDocumentName))
+                .filter(f -> f.getKey().getRecommenderId() == this.recommender.getId().longValue())
+                .map(Map.Entry::getValue)
+                .filter(s -> s.isVisible())
+                .collect(Collectors.toList());
+        
+        try {
+        	ObjectOutputStream objos= new ObjectOutputStream(new FileOutputStream("D://recommender.txt"));
```
Comment (munterkalmsteiner) [inline inception-imls-stringmatch/src/main/java/de/tudarmstadt/ukp/inception/recommendation/imls/stringmatch/StringMatchingRecommender.java]: Hardcoded file and path is not good... I think the intended behaviour is that the user should be prompted by the browser to store or view the exported model file.

File: inception-recommendation-api/src/main/java/de/tudarmstadt/ukp/inception/recommendation/api/recommender/RecommendationEngine.java [hunk 4]
```diff
@@ -152,4 +156,9 @@ protected Feature getIsPredictionFeature(CAS aCas)
     {
         return getPredictedType(aCas).getFeatureByBaseName(FEATURE_NAME_IS_PREDICTION);
     }
+    
+    /**
+     * Method supporting export the recommender object
+     */
+    public abstract void export(RecommenderContext aContext, OutputStream aTarget) throws IOException;
```
Comment (munterkalmsteiner) [inline inception-recommendation-api/src/main/java/de/tudarmstadt/ukp/inception/recommendation/api/recommender/RecommendationEngine.java]: Shouldn't this method not be implemented by all recommenders?

File: inception-imls-stringmatch/src/main/java/de/tudarmstadt/ukp/inception/recommendation/imls/stringmatch/StringMatchingRecommender.java (hunk 3 above)
Comment (reckart) [reply inception-imls-stringmatch/src/main/java/de/tudarmstadt/ukp/inception/recommendation/imls/stringmatch/StringMatchingRecommender.java]: We usually use an `AjaxDownloadLink` when a user should export something:

https://github.com/inception-project/inception/blob/b578054f77891bb152c22bb3a9ee8e3126d31a7c/inception-ui-kb/src/main/java/de/tudarmstadt/ukp/inception/ui/kb/project/AccessSpecificSettingsPanel.java#L342-L343

**Comment by reckart** [comment]
Superseded by #2085.

**Pull request closed without merging**